### ISSUE TITLE
"(1 + 2)" and grammar

### ISSUE THREAD

**Issue by banditpig** [ehsanmok/create-your-own-lang-with-rust#14]
Hi, Thanks for the interesting tutorial! I'm just learning about PEG etc and wondered how to modify the grammar so that 
//assert_eq!(Interpreter::source("(1 + 2)").unwrap() as i32, 3);
will work?

Thanks
Mike

**Comment by PrestonHager** [comment]
[Pest](https://pest.rs) gives a great interactive parser at the bottom of their page. Check it out. The problem with giving the first term wrapped in parentheses is that the grammar tries to parse it as a full Binary Expression first. While this is true, it checks for least recursive first. To fix this, I changed the binary expression to include at least one operator and term. Then an expression can also be a term too (what was originally a binary term with no suffixed operator and term). The following is a fixed version of the grammar code. I also changed an Int to be defined as either negative or non-negative (getting rid of the possibility of having a preceding plus sign).

```
Program = _{ SOI ~ Expr ~ EOF }

Expr = { UnaryExpr | BinaryExpr | Term }

Term = _{Int | "(" ~ Expr ~ ")" }

UnaryExpr = { Operator ~ Term }

BinaryExpr = { Term ~ (Operator ~ Term)+ }

Operator = { "+" | "-" }

Int = @{ "-"? ~ ASCII_DIGIT+ }

WHITESPACE = _{ " " | "\t" }

EOF = _{ EOI | ";" }
```

Also check out pull request https://github.com/ehsanmok/create-your-own-lang-with-rust/pull/16.
